
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.13">
    
    
      
        <title>Use Docker to run PHG - Practical Haplotype Graph Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.85bb2934.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#use-docker-to-run-phg" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Practical Haplotype Graph Documentation" class="md-header__button md-logo" aria-label="Practical Haplotype Graph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Practical Haplotype Graph Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Use Docker to run PHG
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../Home_variantsInGVCFFiles/" class="md-tabs__link">
        Getting Started
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../ImputeWithPHG_extractPangenome/" class="md-tabs__link">
        FAQ
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="https://javadoc.io/doc/net.maizegenetics/phg/latest/index.html" class="md-tabs__link">
      API Reference
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../getting-help/" class="md-tabs__link">
        Where to Ask Questions
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Practical Haplotype Graph Documentation" class="md-nav__button md-logo" aria-label="Practical Haplotype Graph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Practical Haplotype Graph Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
      
      
      
        <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="https://maizegenetics.net" class="md-nav__link">
        Buckler Lab Website
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          Getting Started
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Getting Started
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../Home_variantsInGVCFFiles/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreatePHG_step1CreateDB_loadRef/" class="md-nav__link">
        Step 1: Create Initial PHG with Reference Data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreatePHG_step2AssemblyAndWGSHaplotypes/" class="md-nav__link">
        Step 2: Add Haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImputeWithPHG_main/" class="md-nav__link">
        Step 3: Impute Variants or Haplotypes
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          FAQ
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          FAQ
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImputeWithPHG_extractPangenome/" class="md-nav__link">
        Extract a Pangenome FASTA
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindLikelyParents/" class="md-nav__link">
        Find Likely Parents for Imputation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImputePHG_Simple/" class="md-nav__link">
        PHG Path from a FASTQ
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImputeWithPHG_VCF/" class="md-nav__link">
        Impute Variants from a VCF
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImputeWithPHG_fastq-heterozygous/" class="md-nav__link">
        Impute Variants from a FASTQ
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImputeWithPHG_fastq-homozygous/" class="md-nav__link">
        Impute Variants from a FASTQ (Homozygous)
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../UpdatePHGSchema/" class="md-nav__link">
        Update Existing PHG Database to Latest Schema
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://javadoc.io/doc/net.maizegenetics/phg/latest/index.html" class="md-nav__link">
        API Reference
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          Where to Ask Questions
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Where to Ask Questions
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../getting-help/" class="md-nav__link">
        Biostars
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="use-docker-to-run-phg">Use Docker to run PHG</h1>
<h1 id="quick-start">Quick Start</h1>
<ol>
<li>Download and install Docker following <a href="https://docs.docker.com/get-docker/">instructions</a> on the docker website.</li>
<li>Pull the PHG docker images from Docker Hub.  It is recommended you pull an image based on a specific tag. Knowing which specific PHG image was used can aid with debugging, and make your pipeline more reproducible.  Change the "0.0.40" tag below to match the PHG tag you wish to pull.</li>
</ol>
<pre><code>docker pull maizegenetics/phg:0.0.40
</code></pre>
<p>If on a Cornell CBSU machine, use docker1 as below:</p>
<pre><code>docker1 pull maizegenetics/phg:0.0.40
</code></pre>
<h1 id="details">Details</h1>
<p><a href="../../images/PHGDockerDiagramLarge.png"><img alt="PHGDocker9pt.png" src="../../images/PHGDocker9pt.png" /></a></p>
<p>The PHG is deployed as a Docker image. Docker is an open source platform that wraps applications into an image to be deployed on a host system. Docker containers are instances of a Docker image that may run simultaneously on the same host. See the Docker documentation <a href="https://docs.docker.com">here</a> for details.</p>
<p>The PHG Docker image is built on the GATK image from <a href="https://hub.docker.com">docker hub</a>. To this image several genomic analysis tools are added, including: VIM, BWA, minimap2, jbwa, htslib, bcftools, SAMTools and TASSEL-5 with the PHG jar. In addition, scripts for loading and processing the PHG are added.</p>
<p>To pull the PHG image from docker hub, run the following commands (replace <version tag> with the tagged version you wish to run):</p>
<pre><code>docker pull maizegenetics/phg:&lt;version tag&gt;
</code></pre>
<p>If on a Cornell CBSU machine, use docker1 as below (Replace <version tag> with the tagged version you wish to run):</p>
<pre><code>docker1 pull maizegenetics/phg:&lt;version tag&gt;
</code></pre>
<p>Access to PHG databases in all cases is through a config file that specifies the host:port, user, password, db name, and the database type (sqlite or postgres). An example of the relevant portions of an SQLite config file is below. NOTE: The user and password parameters are not used for SQLite. The sqlite db file must include a path that is relative to its location within the PHG docker scripts. This will be discussed when each script is discussed.</p>
<pre><code>host=localHost
user=sqlite
password=sqlite
DB=/tempFileDir/outputDir/phgTestDB_mapq48.db
DBtype=sqlite
</code></pre>
<p>An example of a postgres config file is below:</p>
<pre><code>host=172.17.0.2:5432\n
user=postgres\n
password=postgres_pwd\n
DB=phgdb\n
DBtype=postgres
</code></pre>
<p>On CBSU NOTE: If unable to pull a docker image from docker hub on a Cornell CBSU machine, check if you are on an enhanced security machine. cbsumm01, 03, 11 and others are. You can see this if you go to basic biohpc reservations pages, and look at the machines you're able to reserve. They will have a note in the description about enhanced security.</p>
<h1 id="kitchen-sink">Kitchen Sink</h1>
<h2 id="why-postgresql"><em>Why PostgreSQL?</em></h2>
<p>PostgreSQL is an industry standard object-relational database system providing reliability and data integrity.  It provides good performance, supports many concurrent users, and allows for user extensions.</p>
<h2 id="why-postgresql-in-a-docker"><em>Why PostgreSQL in a docker?</em></h2>
<p>If you do not have access to Postgres on your server, you can download a containerized version of postgres. To do this execute the following comand, replacing <version> with the version of your choice.</p>
<pre><code>
docker pull postgres:&lt;version&gt;
</code></pre>
<p>Providing a containerized version of PostgreSQL removes the need for the user to intall and maintain their own postgres image.  The database file created and manipulated in the postgres docker can be persisted to a shared volume for processing by multiple applications.</p>
<h2 id="creating-the-phg-postgresql-docker-image"><em>Creating the PHG PostgreSQL docker image</em></h2>
<p>The postgres docker image may be downloaded from docker hub via the following command.  Replace <version> with the postgres version you wish to run.</p>
<pre><code>
  docker pull postgres:&lt;version&gt;
</code></pre>
<p>If running on a Cornell CBSU machine, replace "docker" with "docker1".</p>
<p>If the postgres docker has already been installed and you wish to upgrade, stop and remove any containers associated with the image.  Then remove the existing image using the docker "rmi" command.</p>
<h2 id="connecting-to-a-postgresql-docker-image"><em>Connecting to a PostgreSQL docker image</em></h2>
<p>Connecting to a PHG postgres docker image requires two files:</p>
<ol>
<li>a parameters file</li>
<li>a container commands file.</li>
</ol>
<p>The container commands file reads a user supplied parameters file, then creates a container against the postgres docker image using sourced parameters.  The container is created in detached mode, which allows for "docker exec" commands to be run against it.</p>
<p>Here is an example of a container commands file (container_cmdsPostgres.sh):</p>
<pre><code>#!python

#!/usr/bin/env bash
set -u
set -e
set -x

# read in parameters
source $1

# run the docker in detach mode
docker run -i --detach --name $DOCKER_CONTAINER_NAME -e POSTGRES_PASSWORD=$POSTGRES_PASSWORD -v $POSTGRES_DATA_MOUNT:/var/lib/postgresql/data -p $DOCKER_DB_PORT:5432 $DOCKER_IMAGE_NAME

# start the docker
docker start $DOCKER_CONTAINER_NAME

# PHG scripts may now be run using the postgres db with the host ip address and DOCKER_DB_PORT assigned above
</code></pre>
<p>A sample parameters file, with a description of each parameter, is below.  Among other parameters, the defining of a value for the postgres password helps ensure the security of your data.  Note where each parameter below is used in the container commands file above:</p>
<pre><code>#!python

# This file provides parameters for the container_cmdsPostgresDocker.sh file.
# Edit this file to reflect your configuration

# Port that is mapped to the docker postgres 5432 port.  If running on a Cornell CBSU machine,
# the mapped port must be between 8009-8019.  Those are the only exposed ports.
DOCKER_DB_PORT=&quot;5433&quot;

# Name for the docker container running against the postgres image
CONTAINER_NAME=&quot;phg_postgres_container&quot;

# When running on Cornell CBSU machines, a prefix is appended to any docker container created.
# This prefix consists of the user id followed by &quot;__biohpc_&quot;, e.g. &quot;lcj34__biohpc-&quot;
# This prefix can remain blank if not running on Cornell CBSU machine
CONTAINER_PREFIX=&quot;&quot;
DOCKER_CONTAINER_NAME=${CONTAINER_PREFIX}${CONTAINER_NAME}

# Name of the docker image against which your container will be run
DOCKER_IMAGE_NAME=&quot;postgres:11&quot;

# password for postgres db user.  It is recommended this be changed from the default &quot;postgres&quot;
# The  POSTGRES_PASSWORD must remain consistent when attaching to the same persisted database
# This password should appear in the config file used with other PHG scripts for loading data
# to the phg database.
POSTGRES_PASSWORD=&quot;yoursecretpassword&quot;

# Place to mount the data.  Data will be mapped to internal postgres /var/lib/postgresql/data directory
# This data will persist to the volume specified in POSTGRES_DATA_MOUNT.
POSTGRES_DATA_MOUNT=&quot;/Users/lcj34/postgres_docker/test_postgres_data/data&quot;
</code></pre>
<p>To set up the postgres docker, execute the container commands files (container_cmdsPostgres.sh above) from the command line. The db is now ready for connections. Follow the rules below for defining the host/port in the config file if you would like to connect to this postgres instance from the PHG pipeline.</p>
<h2 id="rules-of-thumb-for-connecting-to-the-postgresql-docker-image-for-phg"><em>Rules of Thumb for Connecting to the PostgreSQL docker image for PHG</em></h2>
<p>In all cases you must ensure the user name/password you used when creating the database initially remains the same across all containers that connect to the PHG postgres docker image, and in all config files used by scripts that access the database.  </p>
<h3 id="connecting-from-an-external-machine">Connecting from an external machine:</h3>
<ul>
<li>your config file must have the internet address of the host where the docker container resides</li>
<li>the port must be the mapped port (value of $DOCKER_DB_PORT) from the "docker run --detach ... command" (see -p = $DOCKER_DB_PORT:5432 in the container and parameters files above)</li>
<li>example:  If the remote machine's ip address is 128.32.5.269 with a mapped port of 5433, the config file would have a host line of "host=128.32.5.269:5433"</li>
</ul>
<h3 id="connecting-from-a-phg-docker-script-on-the-same-host">Connecting from a PHG docker script on the same host:</h3>
<ul>
<li>your config file should have host ip addr of 0.0.0.0:<db_mapped_port></li>
<li>if your machine is using a Docker virtual network, the port would instead be the regular postgres port of 5432</li>
<li>CAVEAT:  This does not always work.  If you are having problems, try using the external machine address, ie host=128.32.5.269:5433 (replacing the IP address and mapped port with your machine's address and mapped port).</li>
</ul>
<h3 id="connecting-from-a-non-docker-program-on-the-same-host">Connecting from a non-docker program on the same host:</h3>
<ul>
<li>your config file should have host ip addr of 0.0.0.0:<db_mapped_port></li>
</ul>
<h3 id="connecting-docker-docker-on-a-cornell-cbsu-machine-both-dockers-on-the-same-machine">Connecting docker-docker on a Cornell CBSU machine (both dockers on the same machine):</h3>
<ul>
<li>your config file should have host ip addr of 0.0.0.0:5432</li>
<li>If the above doesn't work, use the full external address with mapped port.  This takes you out and back in the network which shouldn't be required, but it works if nothing else does.</li>
<li>When a postgres docker run has finished creating and loading a new db, "root" will own all the directories.  In order to run subsequent scripts that mount to this data directory you must "claim" and change permissions on your postgres data directory as below:</li>
</ul>
<pre><code>#!python

&gt; docker1 claim
&gt; chmod -R 777 data  
</code></pre>
<h1 id="troubleshooting">Troubleshooting</h1>
<ol>
<li>Docker can be tricky to get running because it requires root access. If you are having problems, double-check that docker has permissions to access root on your machine.</li>
<li>On CBSU NOTE: If unable to pull a docker image from docker hub on a Cornell CBSU machine, check if you are on an enhanced security machine. cbsumm01, 03, 11 and others are. You can see this if you go to basic biohpc reservations pages, and look at the machines you're able to reserve. They will have a note in the description about enhanced security.</li>
</ol>
<p><a href="../CreatePHG_step0_main/">Return to Step 0 pipeline</a></p>
<p><a href="../../Home_variantsInGVCFFiles/">Return to PHG Version 1 instructions</a></p>
<p><a href="../Home.md">Return to Wiki Home</a></p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "search.suggest", "search.highlight", "search.share"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.b4d07000.min.js"></script>
      
    
  </body>
</html>