
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../CreateHaplotypesFromFastq/">
      
      
        <link rel="next" href="../CreateReferenceIntervals/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.13">
    
    
      
        <title>Overview - Practical Haplotype Graph Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.85bb2934.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#overview" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Practical Haplotype Graph Documentation" class="md-header__button md-logo" aria-label="Practical Haplotype Graph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Practical Haplotype Graph Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Overview
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Practical Haplotype Graph Documentation" class="md-nav__button md-logo" aria-label="Practical Haplotype Graph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Practical Haplotype Graph Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Home_variantTables/" class="md-nav__link">
        How to use the PHG
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Home_variantsInGVCFFiles/" class="md-nav__link">
        How to use the PHG
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Home_version1/" class="md-nav__link">
        Home version1
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../PHG_GUI/" class="md-nav__link">
        Documentation for Genotyping using the PHG within TASSEL GUI
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          PHG 
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          PHG 
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PHG%20/%20BrAPI%20Server%20on%20CBSU/" class="md-nav__link">
         BrAPI Server on CBSU
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" checked>
      
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          Pipeline version1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          Pipeline version1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AssemblyHaplotypesPlugin/" class="md-nav__link">
        AssemblyHaplotypesPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ConfigFile/" class="md-nav__link">
        ConfigFile Documentation #
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateAssemblyAnchorsPlugin/" class="md-nav__link">
        CreateAssemblyAnchorsPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateConsensi/" class="md-nav__link">
        CreateConsensi
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypeKeyFile/" class="md-nav__link">
        CreateHaplotype Key File
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypes/" class="md-nav__link">
        Create Haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypesFromBAM/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypesFromFastq/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Overview
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Overview
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#note-as-of-phg-version-0020-the-keyfile-plugin-parameter-has-been-renamed-wgskeyfile" class="md-nav__link">
    Note, as of PHG version 0.0.20, the keyFile plugin parameter has been renamed wgsKeyFile.
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#more-information-about-the-keyfile-format-is-here" class="md-nav__link">
    More information about the keyfile format is here:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#we-strongly-suggest-you-use-an-external-tool-so-you-know-exactly-what-filters-are-applied-to-your-gvcf-file" class="md-nav__link">
    We strongly suggest you use an external tool so you know exactly what filters are applied to your GVCF file.
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-steps-run-by-createhaplotypesfromgvcf" class="md-nav__link">
    Pipeline Steps run by CreateHaplotypesFromGVCF
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-run-command" class="md-nav__link">
    Example Run Command
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-flags-note-some-parameters-have-both-a-short-and-long-name" class="md-nav__link">
    Command Line Flags (note some parameters have both a short and long name)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sample-script" class="md-nav__link">
    Sample Script:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config-file-parameters" class="md-nav__link">
    Config File Parameters
  </a>
  
    <nav class="md-nav" aria-label="Config File Parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#file-directories" class="md-nav__link">
    File Directories.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tassel-parameters" class="md-nav__link">
    TASSEL parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phg-createhaplotypes-parameters" class="md-nav__link">
    PHG CreateHaplotypes Parameters
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateReferenceIntervals/" class="md-nav__link">
        CreateReferenceIntervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ExportPath/" class="md-nav__link">
        ExportPath
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FastqToHapCountPlugin/" class="md-nav__link">
        FastqToHapCountPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FilterGVCFPlugin/" class="md-nav__link">
        FilterGVCFPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindHaplotypeClustersPlugin/" class="md-nav__link">
        FindHaplotypeClustersPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindPath/" class="md-nav__link">
        FindPath
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindPathKeyFiles/" class="md-nav__link">
        Find Path Key Files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindPathMinimap2/" class="md-nav__link">
        Overview #
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../HapCountBestPathToTextPlugin/" class="md-nav__link">
        HapCountBestPathToTextPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImportHaplotypePathFilePlugin/" class="md-nav__link">
        ImportHaplotypePathFilePlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImportReadMappingToDBPlugin/" class="md-nav__link">
        ImportReadMappingToDBPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../IndexPangenome/" class="md-nav__link">
        Overview #
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadAssemblyAnchors/" class="md-nav__link">
        LoadAssemblyAnchors
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadConsensusAnchorSequencesPlugin/" class="md-nav__link">
        LoadConsensusAnchorSequencesPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadGenomeIntervals/" class="md-nav__link">
        LoadGenomeIntervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadGenomeIntervalsToDBPlugin/" class="md-nav__link">
        LoadGenomeIntervalsToDBPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadHapSequencesToDBPlugin/" class="md-nav__link">
        LoadHapSequencesToDBPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../MergeGVCFPlugin/" class="md-nav__link">
        MergeGVCFPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Minimap2NoDocker/" class="md-nav__link">
        Minimap2NoDocker
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PHG_Liquibase/" class="md-nav__link">
        PHG Liquibase
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PHG_PostgresDocker/" class="md-nav__link">
        PHG PostgresDocker
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ParallelAssemblyAnchorsLoad/" class="md-nav__link">
        ParallelAssemblyAnchorsLoad
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PathsToVCFPlugin/" class="md-nav__link">
        PathsToVCFPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../RunHapCollapsePlugin/" class="md-nav__link">
        RunHapCollapsePlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../SAMToMappingPlugin/" class="md-nav__link">
        SAMToMappingPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../help_biostars/" class="md-nav__link">
        How to ask questions about the PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7_36" >
      
      
      
        <label class="md-nav__link" for="__nav_7_36" id="__nav_7_36_label" tabindex="0">
          DeprecatedScripts
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_36_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7_36">
          <span class="md-nav__icon md-icon"></span>
          DeprecatedScripts
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../DeprecatedScripts/CreateHaplotypes_sh/" class="md-nav__link">
        THIS SCRIPT IS DEPRECATED
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          UserInstructions
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          UserInstructions
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_conda/" class="md-nav__link">
        Use Conda to run PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_docker/" class="md-nav__link">
        Use Docker to run PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_main/" class="md-nav__link">
        CreatePHG step0 main
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_singularity/" class="md-nav__link">
        Use Singularity to run PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1-2_main/" class="md-nav__link">
        Create and Populate a PHG database with haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1CreateDB_loadRef/" class="md-nav__link">
        Create and Populate a PHG database with haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1_bedfile/" class="md-nav__link">
        Create bedfile of intervals for PHG reference ranges
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1_groupGenomeIntervals/" class="md-nav__link">
        Separate genome intervals into different groups
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1_loadGenomeIntervals/" class="md-nav__link">
        Load genome intervals to PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2AssemblyAndWGSHaplotypes/" class="md-nav__link">
        Step 2: Add haplotypes to a database
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_FilterGVCFPlugin/" class="md-nav__link">
        Filter GVCF Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_FilterGVCFSingleFilePlugin/" class="md-nav__link">
        FilterGVCFSingleFilePlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_LoadHaplotypesFromGVCFPluginDetails/" class="md-nav__link">
        LoadHaplotypesFromGVCFPlugin Details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_MAFToGVCFPluginDetails/" class="md-nav__link">
        MAFToGVCFPlugin Details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHaplotypesFromWGS/" class="md-nav__link">
        Quick Start
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHapsFromBAM/" class="md-nav__link">
        Separate genome intervals into groups
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHapsFromFastq/" class="md-nav__link">
        Align WGS fastq files to reference genome
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHapsFromGVCF/" class="md-nav__link">
        Filter GVCF and add variants to database
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_alignAssemblies/" class="md-nav__link">
        Align assemblies to reference genome
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_assemblyAnchorWavePluginDetails/" class="md-nav__link">
        AssemblyMAFFromAnchorWavePlugin Details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_assemblyHaplotypeParameters/" class="md-nav__link">
        Quick Start
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_assemblyViaAnchorwave/" class="md-nav__link">
        Quick Start
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_consensus/" class="md-nav__link">
        Create consensus haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_haplotypeKeyfile/" class="md-nav__link">
        CreateHaplotype Key File
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_taxaGroupTables/" class="md-nav__link">
        Separate taxa  into different groups
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ExampleDatabase/" class="md-nav__link">
        Creating a Simple Example Database
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/FindLikelyParents/" class="md-nav__link">
        Finding Likely Parents for Imputation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/HaplotypeMethod/" class="md-nav__link">
        Haplotype Methods
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputePHG_Simple/" class="md-nav__link">
        I have a fastq file - I want a PHG path
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_VCF/" class="md-nav__link">
        Impute variants from a VCF file
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_compare/" class="md-nav__link">
        Compare PHG variants to existing variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_exportVCF/" class="md-nav__link">
        Export variants as VCF
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_extractPangenome/" class="md-nav__link">
        Extract a pangenome fasta file
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_fastq-heterozygous/" class="md-nav__link">
        Impute variants from any fastq file
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_fastq-homozygous/" class="md-nav__link">
        Impute variants from any fastq file - process for homozygous individuals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_findPathKeyFiles/" class="md-nav__link">
        Find Path Key Files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_main/" class="md-nav__link">
        ImputeWithPHG main
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_paramTuning/" class="md-nav__link">
        Tune pathfinding parameters
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MakeDefaultDirectory/" class="md-nav__link">
        Make Default Directory Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MakeInitialPHGDBPipeline/" class="md-nav__link">
        Make Initial PHG DB Pipeline Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MergeFastqPluginDetailedDocs/" class="md-nav__link">
        MergeFastqPlugin Detailed Documentation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MultisampleBAMPipeline/" class="md-nav__link">
        Step 3a: Create ReadMappings using the MultisampleBAM Pipeline
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/PopulatePHGDBPipeline/" class="md-nav__link">
        Populate PHG DB Pipeline Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/SampleHaplotypeConfigFile/" class="md-nav__link">
        Sample Config file for Loading Haplotypes.
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/UpdatePHGSchema/" class="md-nav__link">
        Update an existing PHG database to latest schema with liquibase
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/help_biostars/" class="md-nav__link">
        How to ask questions about the PHG
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#note-as-of-phg-version-0020-the-keyfile-plugin-parameter-has-been-renamed-wgskeyfile" class="md-nav__link">
    Note, as of PHG version 0.0.20, the keyFile plugin parameter has been renamed wgsKeyFile.
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#more-information-about-the-keyfile-format-is-here" class="md-nav__link">
    More information about the keyfile format is here:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#we-strongly-suggest-you-use-an-external-tool-so-you-know-exactly-what-filters-are-applied-to-your-gvcf-file" class="md-nav__link">
    We strongly suggest you use an external tool so you know exactly what filters are applied to your GVCF file.
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-steps-run-by-createhaplotypesfromgvcf" class="md-nav__link">
    Pipeline Steps run by CreateHaplotypesFromGVCF
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-run-command" class="md-nav__link">
    Example Run Command
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-flags-note-some-parameters-have-both-a-short-and-long-name" class="md-nav__link">
    Command Line Flags (note some parameters have both a short and long name)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sample-script" class="md-nav__link">
    Sample Script:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config-file-parameters" class="md-nav__link">
    Config File Parameters
  </a>
  
    <nav class="md-nav" aria-label="Config File Parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#file-directories" class="md-nav__link">
    File Directories.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tassel-parameters" class="md-nav__link">
    TASSEL parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phg-createhaplotypes-parameters" class="md-nav__link">
    PHG CreateHaplotypes Parameters
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="overview">Overview</h1>
<p>This Groovy Script will create haplotypes originating from a set of filtered GVCF files specified in a <a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/DockerPipeline/CreateHaplotypeKeyFile">keyfile</a>.  This groovy script can be run by itself, but it is also called from <a href="CreateHaplotypesFromBAM">CreateHaplotypesFromBAM.groovy</a> (<a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/src/master/docker/buildFiles/scripts/groovy/CreateHaplotypesFromBAM.groovy">source</a>).  Basically if you already have GVCFs that are filtered to your liking and are ready to be uploaded, you should use this script.</p>
<h3 id="note-as-of-phg-version-0020-the-keyfile-plugin-parameter-has-been-renamed-wgskeyfile">Note, as of PHG version 0.0.20, the keyFile plugin parameter has been renamed wgsKeyFile.</h3>
<h2 id="more-information-about-the-keyfile-format-is-here">More information about the keyfile format is here:</h2>
<p>https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/DockerPipeline/CreateHaplotypeKeyFile.</p>
<p>Note only records with type GVCF will be processed by this plugin.</p>
<p>If you need to filter, either use an external tool like <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> or <a href="https://vcftools.github.io/index.html">vcftools</a>.  Alternatively you can use <a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/DockerPipeline/FilterGVCFPlugin">FilterGVCFSingleFilePlugin</a> in PHG. The Documentation for FilterGVCFPlugin describes the possible filters applied to the GVCF file.  </p>
<p>For Maize, we set mapQ=48, DP_poisson_min=0.01, DP_poisson_max=0.99, GQ_min=50 and filterHets=t. <strong>NOTE THESE ARE FOR MAIZE WGS WITH GOOD COVERAGE &gt;5X.</strong>  These can be used as starting points, but you should not follow these values blindly.  </p>
<h2 id="we-strongly-suggest-you-use-an-external-tool-so-you-know-exactly-what-filters-are-applied-to-your-gvcf-file"><strong>We strongly suggest you use an external tool so you know exactly what filters are applied to your GVCF file.</strong></h2>
<h2 id="pipeline-steps-run-by-createhaplotypesfromgvcf">Pipeline Steps run by CreateHaplotypesFromGVCF</h2>
<ol>
<li>Extract out the Reference Ranges and load the haplotypes to the DB</li>
</ol>
<h2 id="example-run-command">Example Run Command</h2>
<pre><code>#!bash

CreateHaplotypesFromGVCF.groovy -[hd] -config [dbConfigFile]
</code></pre>
<h2 id="command-line-flags-note-some-parameters-have-both-a-short-and-long-name">Command Line Flags (note some parameters have both a short and long name)</h2>
<pre><code>#!bash

[-h, -help] 'Show usage information'
[-d, -description] 'Show information about this Pipeline'
[-config] 'DB Config File(required)'
</code></pre>
<p>This Groovy Script will open up the <a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/DockerPipeline/CreateHaplotypeKeyFile">keyFile</a>(location is specified as the LoadHaplotypesFromGVCFPlugin.keyFile entry in the config file) and will loop over all GVCF records and upload to the DB.  If multiple GVCFs are stored for a given taxon, each GVCF will be treated as an independent haplotype.</p>
<h2 id="sample-script">Sample Script:</h2>
<pre><code>#!bash
#Set these properties
REF_DIR=/workdir/user/DockerTuningTests/InputFiles/Reference/
GVCF_DIR=/workdir/user/DockerTuningTests/InputFiles/
GVCF_DIR_IN_DOCKER=/tempFileDir/data/outputs/gvcfs/
DB=/workdir/user/DockerTuningTests/DockerOutput/phgTestMaizeDB.db
CONFIG_FILE=/workdir/user/DockerTuningTests/DataFolders/LoadRefDataDocker/config.txt
CONFIG_FILE_IN_DOCKER=/tempFileDir/data/config.txt
KEY_FILE=/workdir/user/DockerTuningTests/DataFolders/LoadRefDataDocker/keyfile.txt
KEY_FILE_IN_DOCKER=/tempFileDir/data/keyFile.txt

docker run --name small_seq_test_container --rm \
        -v ${REF_DIR}:/tempFileDir/data/reference/ \
        -v ${GVCF_DIR}:${GVCF_DIR_IN_DOCKER} \
        -v ${DB}:/tempFileDir/outputDir/phgTestMaizeDB.db \
        -v ${CONFIG_FILE}:${CONFIG_FILE_IN_DOCKER} \
        -v ${KEY_FILE}:${KEY_FILE_IN_DOCKER} \
        -t maizegenetics/phg \
        /CreateHaplotypesFromGVCF.groovy \
            -config ${CONFIG_FILE_IN_DOCKER}

</code></pre>
<p>Note the GVCF_DIR_IN_DOCKER needs to also be specified in the config file under the Parameter:LoadHaplotypesFromGVCFPlugin.gvcfDir</p>
<h2 id="config-file-parameters">Config File Parameters</h2>
<p>This CreateHaplotypes Script will need some configuration parameters set in order to work correctly.  These must go in you config.txt file. You must set referenceFasta, LoadHaplotypesFromGVCFPlugin.keyFile, LoadHaplotypesFromGVCFPlugin.gvcfDir, LoadHaplotypesFromGVCFPlugin.referenceFasta, LoadHaplotypesFromGVCFPlugin.haplotypeMethodName.</p>
<h4 id="file-directories">File Directories.</h4>
<ul>
<li>tempFileDir=/tempFileDir/data/bam/temp/<ul>
<li>The Temp file directory.  This Location is used to make any intermediate files when processing.  For this script, only the BED File will be extracted containing the Focused Reference Ranges setup in a previous step.</li>
</ul>
</li>
</ul>
<h4 id="tassel-parameters">TASSEL parameters</h4>
<ul>
<li>Xmx=10G<ul>
<li>Max Java Heap Space used when running TASSEL code.</li>
</ul>
</li>
<li>tasselLocation=/tassel-5-standalone/run_pipeline.pl<ul>
<li>Location of TASSEL on machine.  If using PHG docker this is the correct location</li>
</ul>
</li>
</ul>
<h4 id="phg-createhaplotypes-parameters">PHG CreateHaplotypes Parameters</h4>
<ul>
<li>referenceFasta<ul>
<li>Reference fasta file location.  <strong>This is Required.</strong> Note, if using Docker, this needs to be the Docker specific path to the reference file.  You will need to mount these files to the Docker in order for it to work.</li>
</ul>
</li>
<li>LoadHaplotypesFromGVCFPlugin.wgsKeyFile<ul>
<li>Location of the <a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/DockerPipeline/CreateHaplotypeKeyFile">keyfile</a>. <strong>This is Required.</strong> Note, if using Docker, this needs to be the Docker specific path to the keyfile.  You will need to mount this file to the Docker in order for it to work.  Note before PHG version 0.0.20, this parameter was named keyFile.</li>
</ul>
</li>
<li>LoadHaplotypesFromGVCFPlugin.gvcfDir<ul>
<li>Directory of the GVCF files you wish to upload. <strong>This is Required.</strong> Note, if using Docker, this needs to be the Docker specific path to the gvcfs.  You will need to mount these files to the Docker in order for it to work. </li>
</ul>
</li>
<li>LoadHaplotypesFromGVCFPlugin.referenceFasta<ul>
<li>This should be set to the same location as referenceFasta. <strong>This is Required.</strong></li>
</ul>
</li>
<li>LoadHaplotypesFromGVCFPlugin.haplotypeMethodName=GATK_PIPELINE<ul>
<li>This is the method name which you are going to write to the DB.  <strong>This is Required.</strong> If you attempt to upload the same gvcfs and do not change this an error will be thrown.</li>
</ul>
</li>
<li>LoadHaplotypesFromGVCFPlugin.haplotypeMethodDescription<ul>
<li>This is the description for the haplotypeMethodName.  <strong>This is Required.</strong></li>
</ul>
</li>
<li>extendedWindowSize=1000<ul>
<li>This Script will extract out the Focused reference ranges from the Database into a BED file.  When doing the BAM file filtering, we allow the user to add flanking regions around each of the regions.</li>
</ul>
</li>
<li>numThreads=10<ul>
<li>Number of threads requested to run HaplotypeCaller.  </li>
</ul>
</li>
<li>refRangeMethods=refRegionGroup<ul>
<li>This is used to extract a BED file out of the DB before the GVCF file is processed.  The BED file is then used to extract out regions of the GVCF used to become the haplotypes.  Typically, refRegionGroup refers to the anchor Reference ranges.  If "refRegionGroup,refInterRegionGroup" is used it will create a BED file representing both anchors and inter anchors.  <strong>We strongly suggest not setting this parameter in the Config File</strong></li>
</ul>
</li>
</ul>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.b4d07000.min.js"></script>
      
    
  </body>
</html>