
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../MergeGVCFPlugin/">
      
      
        <link rel="next" href="../PHG_Liquibase/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.13">
    
    
      
        <title>Minimap2NoDocker - Practical Haplotype Graph Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.85bb2934.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#prerequisites" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Practical Haplotype Graph Documentation" class="md-header__button md-logo" aria-label="Practical Haplotype Graph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Practical Haplotype Graph Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Minimap2NoDocker
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Practical Haplotype Graph Documentation" class="md-nav__button md-logo" aria-label="Practical Haplotype Graph Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Practical Haplotype Graph Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Home_variantTables/" class="md-nav__link">
        How to use the PHG
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Home_variantsInGVCFFiles/" class="md-nav__link">
        How to use the PHG
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Home_version1/" class="md-nav__link">
        Home version1
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../PHG_GUI/" class="md-nav__link">
        Documentation for Genotyping using the PHG within TASSEL GUI
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          PHG 
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          PHG 
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PHG%20/%20BrAPI%20Server%20on%20CBSU/" class="md-nav__link">
         BrAPI Server on CBSU
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" checked>
      
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          Pipeline version1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          Pipeline version1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../AssemblyHaplotypesPlugin/" class="md-nav__link">
        AssemblyHaplotypesPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ConfigFile/" class="md-nav__link">
        ConfigFile Documentation #
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateAssemblyAnchorsPlugin/" class="md-nav__link">
        CreateAssemblyAnchorsPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateConsensi/" class="md-nav__link">
        CreateConsensi
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypeKeyFile/" class="md-nav__link">
        CreateHaplotype Key File
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypes/" class="md-nav__link">
        Create Haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypesFromBAM/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypesFromFastq/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateHaplotypesFromGVCF/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CreateReferenceIntervals/" class="md-nav__link">
        CreateReferenceIntervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ExportPath/" class="md-nav__link">
        ExportPath
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FastqToHapCountPlugin/" class="md-nav__link">
        FastqToHapCountPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FilterGVCFPlugin/" class="md-nav__link">
        FilterGVCFPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindHaplotypeClustersPlugin/" class="md-nav__link">
        FindHaplotypeClustersPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindPath/" class="md-nav__link">
        FindPath
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindPathKeyFiles/" class="md-nav__link">
        Find Path Key Files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../FindPathMinimap2/" class="md-nav__link">
        Overview #
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../HapCountBestPathToTextPlugin/" class="md-nav__link">
        HapCountBestPathToTextPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImportHaplotypePathFilePlugin/" class="md-nav__link">
        ImportHaplotypePathFilePlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ImportReadMappingToDBPlugin/" class="md-nav__link">
        ImportReadMappingToDBPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../IndexPangenome/" class="md-nav__link">
        Overview #
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadAssemblyAnchors/" class="md-nav__link">
        LoadAssemblyAnchors
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadConsensusAnchorSequencesPlugin/" class="md-nav__link">
        LoadConsensusAnchorSequencesPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadGenomeIntervals/" class="md-nav__link">
        LoadGenomeIntervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadGenomeIntervalsToDBPlugin/" class="md-nav__link">
        LoadGenomeIntervalsToDBPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../LoadHapSequencesToDBPlugin/" class="md-nav__link">
        LoadHapSequencesToDBPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../MergeGVCFPlugin/" class="md-nav__link">
        MergeGVCFPlugin
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Minimap2NoDocker
      </a>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PHG_Liquibase/" class="md-nav__link">
        PHG Liquibase
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PHG_PostgresDocker/" class="md-nav__link">
        PHG PostgresDocker
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../ParallelAssemblyAnchorsLoad/" class="md-nav__link">
        ParallelAssemblyAnchorsLoad
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../PathsToVCFPlugin/" class="md-nav__link">
        PathsToVCFPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../RunHapCollapsePlugin/" class="md-nav__link">
        RunHapCollapsePlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../SAMToMappingPlugin/" class="md-nav__link">
        SAMToMappingPlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../help_biostars/" class="md-nav__link">
        How to ask questions about the PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7_36" >
      
      
      
        <label class="md-nav__link" for="__nav_7_36" id="__nav_7_36_label" tabindex="0">
          DeprecatedScripts
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_7_36_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7_36">
          <span class="md-nav__icon md-icon"></span>
          DeprecatedScripts
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../DeprecatedScripts/CreateHaplotypes_sh/" class="md-nav__link">
        THIS SCRIPT IS DEPRECATED
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          UserInstructions
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          UserInstructions
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_conda/" class="md-nav__link">
        Use Conda to run PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_docker/" class="md-nav__link">
        Use Docker to run PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_main/" class="md-nav__link">
        CreatePHG step0 main
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step0_singularity/" class="md-nav__link">
        Use Singularity to run PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1-2_main/" class="md-nav__link">
        Create and Populate a PHG database with haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1CreateDB_loadRef/" class="md-nav__link">
        Create and Populate a PHG database with haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1_bedfile/" class="md-nav__link">
        Create bedfile of intervals for PHG reference ranges
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1_groupGenomeIntervals/" class="md-nav__link">
        Separate genome intervals into different groups
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step1_loadGenomeIntervals/" class="md-nav__link">
        Load genome intervals to PHG
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2AssemblyAndWGSHaplotypes/" class="md-nav__link">
        Step 2: Add haplotypes to a database
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_FilterGVCFPlugin/" class="md-nav__link">
        Filter GVCF Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_FilterGVCFSingleFilePlugin/" class="md-nav__link">
        FilterGVCFSingleFilePlugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_LoadHaplotypesFromGVCFPluginDetails/" class="md-nav__link">
        LoadHaplotypesFromGVCFPlugin Details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_MAFToGVCFPluginDetails/" class="md-nav__link">
        MAFToGVCFPlugin Details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHaplotypesFromWGS/" class="md-nav__link">
        Quick Start
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHapsFromBAM/" class="md-nav__link">
        Separate genome intervals into groups
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHapsFromFastq/" class="md-nav__link">
        Align WGS fastq files to reference genome
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_addHapsFromGVCF/" class="md-nav__link">
        Filter GVCF and add variants to database
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_alignAssemblies/" class="md-nav__link">
        Align assemblies to reference genome
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_assemblyAnchorWavePluginDetails/" class="md-nav__link">
        AssemblyMAFFromAnchorWavePlugin Details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_assemblyHaplotypeParameters/" class="md-nav__link">
        Quick Start
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_assemblyViaAnchorwave/" class="md-nav__link">
        Quick Start
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_consensus/" class="md-nav__link">
        Create consensus haplotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_haplotypeKeyfile/" class="md-nav__link">
        CreateHaplotype Key File
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/CreatePHG_step2_taxaGroupTables/" class="md-nav__link">
        Separate taxa  into different groups
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ExampleDatabase/" class="md-nav__link">
        Creating a Simple Example Database
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/FindLikelyParents/" class="md-nav__link">
        Finding Likely Parents for Imputation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/HaplotypeMethod/" class="md-nav__link">
        Haplotype Methods
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputePHG_Simple/" class="md-nav__link">
        I have a fastq file - I want a PHG path
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_VCF/" class="md-nav__link">
        Impute variants from a VCF file
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_compare/" class="md-nav__link">
        Compare PHG variants to existing variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_exportVCF/" class="md-nav__link">
        Export variants as VCF
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_extractPangenome/" class="md-nav__link">
        Extract a pangenome fasta file
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_fastq-heterozygous/" class="md-nav__link">
        Impute variants from any fastq file
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_fastq-homozygous/" class="md-nav__link">
        Impute variants from any fastq file - process for homozygous individuals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_findPathKeyFiles/" class="md-nav__link">
        Find Path Key Files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_main/" class="md-nav__link">
        ImputeWithPHG main
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/ImputeWithPHG_paramTuning/" class="md-nav__link">
        Tune pathfinding parameters
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MakeDefaultDirectory/" class="md-nav__link">
        Make Default Directory Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MakeInitialPHGDBPipeline/" class="md-nav__link">
        Make Initial PHG DB Pipeline Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MergeFastqPluginDetailedDocs/" class="md-nav__link">
        MergeFastqPlugin Detailed Documentation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/MultisampleBAMPipeline/" class="md-nav__link">
        Step 3a: Create ReadMappings using the MultisampleBAM Pipeline
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/PopulatePHGDBPipeline/" class="md-nav__link">
        Populate PHG DB Pipeline Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/SampleHaplotypeConfigFile/" class="md-nav__link">
        Sample Config file for Loading Haplotypes.
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/UpdatePHGSchema/" class="md-nav__link">
        Update an existing PHG database to latest schema with liquibase
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../UserInstructions/help_biostars/" class="md-nav__link">
        How to ask questions about the PHG
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<p>Steps to run Minimap2BasedPipeline without needing a docker.</p>
<h1 id="prerequisites">Prerequisites:</h1>
<ul>
<li>A Database filled with haplotypes</li>
<li>The most recent tassel-5-standalone from bitbucket</li>
<li>The most recent phg.jar built from practicalhaplotypegraph project.  Put this in tassel-5-standalone/lib/</li>
<li>The most recent version of minimap2 downloaded and installed.</li>
</ul>
<h1 id="step-1-pull-out-the-fasta-files">Step 1: pull out the fasta files</h1>
<pre><code>#!bash

time ./tassel-5-standalone/run_pipeline.pl -Xmx100g -debug -HaplotypeGraphBuilderPlugin -configFile [CONFIG_FILE_NAME] -methods [HAPLOTYPE_METHOD_NAME] -includeVariantContexts false -endPlugin -WriteFastaFromGraphPlugin -outputFile [OUTPUT_FASTA_NAME] -endPlugin &gt; [LOG_FILE]
</code></pre>
<h2 id="parameters">Parameters:</h2>
<ul>
<li>CONFIG_FILE_NAME - The name of the config file you are wanting to run through the pipeline.  This is used to figure out what DB to access.</li>
<li>HAPLOTYPE_METHOD_NAME - The name of the Haplotype Methods used when populating the DB.  </li>
<li>OUTPUT_FASTA_NAME - The name of the output fasta file you are creating by running this step.</li>
<li>LOG_FILE - file to write the logs to.</li>
</ul>
<h1 id="step-2-index-the-pangenome-fasta-using-minimap2-30-minutes-for-20-assemblies-with-75k-anchors">Step 2: Index the pangenome fasta using minimap2 (~30 minutes for 20 assemblies with 75k anchors)</h1>
<pre><code>#!bash
time minimap2/minimap2 -d [OUTPUT_MINIMAP_INDEX] -k 21 -w 11 -I 90G [PANGENOME_FASTA_FILE] 

</code></pre>
<h2 id="parameters_1">Parameters:</h2>
<ul>
<li>OUTPUT_MINIMAP_INDEX - output file name for the minimap2 index file.  This should have a .mmi extension</li>
<li>-k minimap2 parameter for kmer-size.  We Tested 21 to be the best for maize.</li>
<li>-w minimap2 parameter for window size.  We tested 11 to be the best for maize.</li>
<li>-I - minimap2 parameter may need to be dropped if running into memory issues.  This should be above the number of basepairs in the entire pangenome.  If it is too low it will batch the index and cause the mapping to be incorrect.</li>
<li>PANGENOME_FASTA_FILE - the output from the last step. </li>
</ul>
<h1 id="step-3-run-fastqdirtomappingpluginthis-one-is-the-slowest-step-highly-dependent-on-read-coverage">Step 3: Run FastqDirToMappingPlugin(This one is the slowest step.  Highly dependent on read coverage.)</h1>
<pre><code>#!bash

time ./tassel-5-standalone/run_pipeline.pl -debug -Xmx200G -HaplotypeGraphBuilderPlugin -configFile [CONFIG_FILE] -methods [HAPLOTYPE_METHOD_NAME] -includeVariantContexts false -includeSequences false  -endPlugin -FastqDirToMappingPlugin -minimap2IndexFile [MINIMAP_INDEX_FILE] -fastqDir [READ_DIRECTORY]/ -mappingFileDir [OUTPUT_MAPPING_DIR]/ -paired [PAIRED] -endPlugin &gt; [LOG_FILE]
</code></pre>
<h2 id="parameters_2">Parameters:</h2>
<ul>
<li>CONFIG_FILE_NAME - The name of the config file you are wanting to run through the pipeline.  This is used to figure out what DB to access.</li>
<li>HAPLOTYPE_METHOD_NAME - The name of the Haplotype Methods used when populating the DB.  </li>
<li>MINIMAP_INDEX_FILE - Minimap2 index file made in previous step</li>
<li>READ_DIRECTORY - directory holding all the reads you wish to align.  Make sure to have a "/" at the end.</li>
<li>OUTPUT_MAPPING_DIR - directory where all the mapping files are stored. Make sure to have a "/" at the end. This will go away likely by fall 2019.</li>
<li>PAIRED - either "true" or "false". If true, it will attempt to pair up reads based on file name and use the pair to align.  For this to work correctly the file names need to be in the form B73_1.fq and B73_2.fq or even B73_R1.fq and B73_R2.fq.  Basically everything but the last token separated by "_" must be the same for the pair.  If false it will map each file independently and make a mapping file for each.  False is generally used for GBS.</li>
</ul>
<h1 id="step-4-find-path">Step 4: Find Path</h1>
<pre><code>#!bash

time ./tassel-5-standalone/run_pipeline.pl -debug -Xmx240g -HaplotypeGraphBuilderPlugin -configFile [CONFIG_FILE_NAME] -methods [HAPLOTYPE_METHOD_NAME],[REF_RANGE_METHOD_NAME] -includeVariantContexts false -includeSequences false -endPlugin -HapCountBestPathToTextPlugin -configFile [CONFIG_FILE_NAME] -inclusionFileDir [MAPPING_DIR]/ -outputDir [OUTPUT_PATH_DIR]/ -hapCountMethod [HAP_COUNT_METHOD_NAME] -pMethod [PATH_METHOD_NAME] -endPlugin &gt; [LOG_FILE] 2&gt;&amp;
</code></pre>
<h2 id="parameters_3">Parameters:</h2>
<ul>
<li>CONFIG_FILE_NAME - The name of the config file you are wanting to run through the pipeline.  This is used to figure out what DB to access.</li>
<li>HAPLOTYPE_METHOD_NAME - The name of the Haplotype Methods used when populating the DB.  </li>
<li>REF_RANGE_METHOD_NAME - Name of the reference ranges you wish to run on.  It is highly suggested you use "refRegionGroup" as this is the genic regions.  This approach does not work very well with the intergenes.</li>
<li>MAPPING_DIR - Name of the mapping directory output from the previous step.
OUTPUT_PATH_DIR - Output directory for the path output files.</li>
<li>HAP_COUNT_METHOD_NAME - name of the hap counts stored in the DB.  Just put a dummy value here for now.  It will be used in the future.</li>
<li>PATH_METHOD_NAME - name of the path method to be stored in the DB.  Just put a dummy value here for now. It will be used in the future.</li>
</ul>
<p>Depending on what you are attempting to do, you may need to tweak the config file parameters.  In particular:</p>
<ul>
<li>minReads - this will control the amount of imputation the HMM will do.  If it is set to 0 the HMM will attempt to find the path through all reference ranges even if we did not have any reads supporting those haplotypes.  If set to 1, we require at least one Read hitting haplotypes in the reference range.</li>
<li>splitTaxa - should probably set this to true when using assemblies.  Assemblies tend to skip over some reference regions which can bottle neck the HMM.  By doing splitTaxa=true, this is solved.</li>
</ul>
<p>Here are the parameters we typically use:</p>
<pre><code>#!
maxNodesPerRange=30
minTaxaPerRange=1
minReads=0
maxReadsPerKB=1000
minTransitionProb=0.001
probReadMappedCorrectly=0.99
emissionMethod=allCounts
splitTaxa=true
useBF=false

</code></pre>
<h1 id="step-5-create-vcfoptional">Step 5: Create VCF(Optional)</h1>
<pre><code>#!bash
time ./tassel-5-standalone/run_pipeline.pl -debug -Xmx100g -HaplotypeGraphBuilderPlugin -configFile [CONFIG_FILE_NAME] -methods [HAPLOTYPE_METHOD_NAME],[REF_RANGE_METHOD_NAME] -includeVariantContexts true -endPlugin -ImportHaplotypePathFilePlugin -inputFileDirectory [PATH_DIR_NAME]/ -endPlugin -PathsToVCFPlugin -positionVCF [KNOWN_SITES_VCF] -ref [REFERENCE_FILE] -outputFile [OUTPUT_VCF_FILE] -endPlugin &gt; [LOG_FILE] 2&gt;&amp;1

</code></pre>
<h2 id="parameters_4">Parameters:</h2>
<ul>
<li>CONFIG_FILE_NAME - The name of the config file you are wanting to run through the pipeline.  This is used to figure out what DB to access.</li>
<li>HAPLOTYPE_METHOD_NAME - The name of the Haplotype Methods used when populating the DB.  </li>
<li>REF_RANGE_METHOD_NAME - Name of the reference ranges you wish to run on.  It is highly suggested you use "refRegionGroup" as this is the genic regions.  This approach does not work very well with the intergenes.</li>
<li>PATH_DIR_NAME - Directory holding the paths. This was filled in the FindPaths step.</li>
<li>KNOWN_SITES_VCF - Optional parameter to force certain sites to be exported.  If you have a known SNP set which you need all the sites to compare, pass that SNP set in as a vcf file.  If not, remove the -positionVCF flag as well.</li>
<li>REFERENCE_FILE - reference fasta used when creating the graph.</li>
<li>OUTPUT_VCF_FILE - name of the output vcf file.</li>
</ul>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.b4d07000.min.js"></script>
      
    
  </body>
</html>